server/src/main/java/org/elasticsearch/index/codec/tsdb/ES97TSDBDocValuesConsumer.java

```diff
@@ -0,0 +1,213 @@
+ */
+
+package org.elasticsearch.index.codec.tsdb;
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.codecs.DocValuesConsumer;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.codecs.lucene90.IndexedDISI;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.EmptyDocValuesProducer;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.store.ByteBuffersDataOutput;
+import org.apache.lucene.store.ByteBuffersIndexOutput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.packed.DirectMonotonicWriter;
+import org.elasticsearch.core.IOUtils;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+final class ES97TSDBDocValuesConsumer extends DocValuesConsumer {
+
+    IndexOutput data, meta;
+    final int maxDoc;
+    private final int numericBlockShift;
+    private final int numericBlockSize;
+    private final int directMonotonicBlockShift;
+
+    ES97TSDBDocValuesConsumer(
+        SegmentWriteState state,
+        String dataCodec,
+        String dataExtension,
+        String metaCodec,
+        String metaExtension,
+        int numericBlockShift,
+        int numericBlockSize,
+        int directMonotonicBlockShift
+    ) throws IOException {
+        boolean success = false;
+        this.numericBlockShift = numericBlockShift;
+        this.numericBlockSize = numericBlockSize;
+        this.directMonotonicBlockShift = directMonotonicBlockShift;
+        try {
+            final String dataName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, dataExtension);
+            data = state.directory.createOutput(dataName, state.context);
+            CodecUtil.writeIndexHeader(
+                data,
+                dataCodec,
+                ES97TSDBDocValuesFormat.VERSION_CURRENT,
+                state.segmentInfo.getId(),
+                state.segmentSuffix
+            );
+            String metaName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, metaExtension);
+            meta = state.directory.createOutput(metaName, state.context);
+            CodecUtil.writeIndexHeader(
+                meta,
+                metaCodec,
+                ES97TSDBDocValuesFormat.VERSION_CURRENT,
+                state.segmentInfo.getId(),
+                state.segmentSuffix
+            );
+            maxDoc = state.segmentInfo.maxDoc();
+            success = true;
+        } finally {
+            if (!success) {
+                IOUtils.closeWhileHandlingException(this);
+            }
+        }
+    }
+
+    @Override
+    public void addNumericField(FieldInfo field, DocValuesProducer valuesProducer) throws IOException {
+        meta.writeInt(field.number);
```

Comment:
NOTE: here I am not writing the type of field (NUMERIC) because this only works for numeric fields and I spare some space not writing a value that would always be the same. I am wondering, anyway, if not writing this might have an impact in terms of BWC incase we later introduce compression for other fields.

Reply:
I would suggest keeping it: we are talking about a single integer per segment per field, so the impact on space efficiency is null, and we could more easily add support for other types of doc values later on.